File: Linkedlist/ts/reverseList.ts
export class ListNode<T> {
  constructor(public val: T, public next: ListNode<T> | null = null) {}
}

export function reverseList<T>(head: ListNode<T> | null): ListNode<T> | null {
  let prev: ListNode<T> | null = null;
  let current = head;

  while (current !== null) {
    const next = current.next;
    current.next = prev;
    prev = current;
    current = next;
  }

  return prev;
}

File: Linkedlist/java/reverseList.java
class ReverseList {
  public static ListNode reverseList(ListNode head) {
    ListNode prev = null;
    ListNode current = head;

    while (current != null) {
      ListNode next = current.next;
      current.next = prev;
      prev = current;
      current = next;
    }

    return prev;
  }

  public static void run() {
    // intentionally empty
  }
}
